File: src/projects.js
import { readFileSync, writeFileSync, mkdirSync, existsSync, unlinkSync, readdirSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';

const PROJECTS_DIR = join(homedir(), '.relay-cli', 'projects');
const ACTIVE_FILE = join(homedir(), '.relay-cli', 'active_project');

function ensureDir() {
  mkdirSync(PROJECTS_DIR, { recursive: true });
}

function projectPath(name) {
  return join(PROJECTS_DIR, `${name}.json`);
}

export function createProject(name, config) {
  ensureDir();
  const p = projectPath(name);
  writeFileSync(p, JSON.stringify(config, null, 2), 'utf8');
  return { name, path: p, ...config };
}

export function getProject(name) {
  const p = projectPath(name);
  if (!existsSync(p)) return null;
  const data = JSON.parse(readFileSync(p, 'utf8'));
  data.name = name;
  return data;
}

export function listProjects() {
  ensureDir();
  const active = getActiveProjectName();
  return readdirSync(PROJECTS_DIR)
    .filter(f => f.endsWith('.json'))
    .sort()
    .map(f => {
      const name = f.replace('.json', '');
      const data = JSON.parse(readFileSync(join(PROJECTS_DIR, f), 'utf8'));
      data.name = name;
      data.active = name === active;
      return data;
    });
}

export function deleteProject(name) {
  const p = projectPath(name);
  if (!existsSync(p)) return false;
  unlinkSync(p);
  if (getActiveProjectName() === name && existsSync(ACTIVE_FILE)) {
    unlinkSync(ACTIVE_FILE);
  }
  return true;
}

export function setActiveProject(name) {
  if (!existsSync(projectPath(name))) {
    throw new Error(`Project '${name}' does not exist.`);
  }
  mkdirSync(join(homedir(), '.relay-cli'), { recursive: true });
  writeFileSync(ACTIVE_FILE, name, 'utf8');
}

export function getActiveProjectName() {
  if (!existsSync(ACTIVE_FILE)) return null;
  const name = readFileSync(ACTIVE_FILE, 'utf8').trim();
  if (existsSync(projectPath(name))) return name;
  return null;
}

export function getActiveProject() {
  const name = getActiveProjectName();
  return name ? getProject(name) : null;
}

File: relay/projects.py
import json
from pathlib import Path

PROJECTS_DIR = Path.home() / ".relay-cli" / "projects"
ACTIVE_FILE = Path.home() / ".relay-cli" / "active_project"


def _ensure_dir():
    PROJECTS_DIR.mkdir(parents=True, exist_ok=True)


def _project_path(name: str) -> Path:
    return PROJECTS_DIR / f"{name}.json"


def create_project(name: str, config: dict) -> dict:
    _ensure_dir()
    path = _project_path(name)
    path.write_text(json.dumps(config, indent=2), encoding="utf-8")
    return {"name": name, "path": str(path), **config}


def get_project(name: str) -> dict | None:
    path = _project_path(name)
    if not path.exists():
        return None
    data = json.loads(path.read_text(encoding="utf-8"))
    data["name"] = name
    return data


def list_projects() -> list[dict]:
    _ensure_dir()
    projects = []
    active = get_active_project_name()
    for f in sorted(PROJECTS_DIR.glob("*.json")):
        data = json.loads(f.read_text(encoding="utf-8"))
        data["name"] = f.stem
        data["active"] = f.stem == active
        projects.append(data)
    return projects


def delete_project(name: str) -> bool:
    path = _project_path(name)
    if path.exists():
        path.unlink()
        if get_active_project_name() == name:
            ACTIVE_FILE.unlink(missing_ok=True)
        return True
    return False


def set_active_project(name: str):
    if not _project_path(name).exists():
        raise RuntimeError(f"Project '{name}' does not exist.")
    ACTIVE_FILE.parent.mkdir(parents=True, exist_ok=True)
    ACTIVE_FILE.write_text(name, encoding="utf-8")


def get_active_project_name() -> str | None:
    if ACTIVE_FILE.exists():
        name = ACTIVE_FILE.read_text(encoding="utf-8").strip()
        if _project_path(name).exists():
            return name
    return None


def get_active_project() -> dict | None:
    name = get_active_project_name()
    if name:
        return get_project(name)
    return None
